File: backend/app/database/models/membership.py
from datetime import datetime

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Index, Integer, String, func, text
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class Membership(Base):
    __tablename__ = "memberships"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    organization_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("organizations.id", ondelete="CASCADE"), index=True, nullable=False
    )
    role_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("roles.id", ondelete="CASCADE"), index=True, nullable=False
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="ACTIVE", server_default=text("'ACTIVE'"))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    __table_args__ = (
        CheckConstraint("status IN ('ACTIVE', 'SUSPENDED', 'INVITED')", name="ck_membership_status"),
        Index(
            "idx_unique_active_membership",
            "user_id",
            "organization_id",
            unique=True,
            postgresql_where=text("status = 'ACTIVE'"),
        ),
    )

File: frontend/src/types/membership.ts
export type MembershipStatus = 'ACTIVE' | 'SUSPENDED' | 'INVITED'

export interface Membership {
  id: number
  organization_id: number
  user_id: number
  role_id: number
  status: MembershipStatus
  user_name: string
  user_email: string
  role_name: string
  created_at: string
  updated_at: string
}

export interface MembershipCreate {
  user_id: number
  role_id: number
}

export interface MembershipUpdate {
  role_id?: number
  status?: MembershipStatus
}
